### PR TITLE
create DuplicateUrlDiscarderPipeline

### PR DESCRIPTION
**Pull request by BurnzZ**

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -146,20 +146,49 @@ non-universal rules that match the URL, the universal ones are applied.
...
-    ]
+      DUD_LOAD_RULE_PATHS = [
+          "/home/user/project/custom_rules1.json",
+      ]
+
+  The default value of this setting is empty. However, if the package
+  `duplicate-url-discarder-rules`_ is installed and ``DUD_LOAD_RULE_PATHS``
+  has been left empty, the rules in the said package is automatically used.
+
+* ``DUD_ATTRIBUTES_PER_ITEM``: it's a mapping of either a type *(or its import path)*
```
Comment (Gallaecio) [inline README.rst:164]:
```suggestion
* ``DUD_ATTRIBUTES_PER_ITEM``: it's a mapping of a type *(or its import path)*
```

File: duplicate_url_discarder/_addon.py [hunk 2]
```diff
@@ -14,3 +30,9 @@ def update_settings(self, settings: BaseSettings) -> None:
             current_fpr,
             "addon",
         )
+        _setdefault(
+            settings,
+            "ITEM_PIPELINES",
+            "duplicate_url_discarder.DuplicateUrlDiscarderPipeline",
```
Comment (Gallaecio) [inline duplicate_url_discarder/_addon.py:36]: `_setdefault` expects a class, not an import path.

File: duplicate_url_discarder/utils.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        new_mapping[cls] = values
+    return new_mapping
+
+
+def item_signature(item: ItemAdapter, item_attributes: Iterable[str]) -> int:
+    try:
+        values = [str(item.get(attrib)) for attrib in item_attributes]
+    except AttributeError:
+        raise ValueError(f"Got type {type(item)} but expected ItemAdapter.")
+    return hash("|".join(values))
```
Comment (Gallaecio) [inline duplicate_url_discarder/utils.py:25]: Values may contain «|». Although it is unlikely to cause an issue , and maybe the performance `json.dumps(values)` is significantly worse.

File: tests/test_pipelines.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+@ensureDeferred
+async def test_duplicate_url_discarder_pipeline_no_addon(caplog, httpserver) -> None:
+    caplog.set_level("INFO")
+
+    httpserver.expect_request("")
+    start_url = httpserver.url_for("")
+
+    class FakeSpider(Spider):
+        name = "fake_spider"
+        start_urls = [start_url]
```
Comment (Gallaecio) [inline tests/test_pipelines.py:18]: If you don’t care about the HTTP handled being used, you could use `data:,` instead, no need for an HTTP server, and the test should be faster.

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    adapter = ItemAdapter(item)
+
+    assert item_signature(adapter, ["name"]) == hash(value)
+
+    exception_text = (
+        "Got type <class 'tests.test_utils.test_item_signature.<locals>.FakeItem'> "
+        "but expected ItemAdapter."
+    )
+    with pytest.raises(ValueError, match=exception_text):
+        item_signature(item, ["name"])  # type: ignore
```
Comment (Gallaecio) [inline tests/test_utils.py:45]: If we keep the current implementation, maybe we could add an xfail test for:

```
item1 = {"a": "a|", "b": "b"}
item2 = {"a": "a", "b": "|b"}
fields = ["a", "b"]
signature = "a||b"
```

File: duplicate_url_discarder/_addon.py (hunk 2 above)
Comment (BurnzZ) [reply duplicate_url_discarder/_addon.py]: Missed that. Thanks for spotting! https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/20b224a7683ccedf26abefdb13fa5a440cac952e

File: tests/test_pipelines.py (hunk 4 above)
Comment (BurnzZ) [reply tests/test_pipelines.py]: TIL. Thanks super handy. Thanks for raising it! https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/c12a803bb85578662a639f2ec68968d2123c8574

File: tests/test_utils.py (hunk 5 above)
Comment (BurnzZ) [reply tests/test_utils.py]: Very interesting edge case! ✨ 

Let's go with adding the attrib name _(or dict key)_ in the hash calculation. As this would mean that the left-hand value is guaranteed to be unique since it's being used as a dict-key or an attribute name. https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/afd2a0f031fa33da71fa834fe0b0c61f4f87d212

File: duplicate_url_discarder/utils.py (hunk 3 above)
Comment (BurnzZ) [reply duplicate_url_discarder/utils.py:25]: Updated with https://github.com/zytedata/duplicate-url-discarder/pull/17#discussion_r1680915094.

Yeah `json.dumps()` has a slightly lower performance than this approach.

File: tests/test_utils.py (hunk 5 above)
Comment (Gallaecio) [reply tests/test_utils.py]: Then:

```
item1 = {"a": "a|b:b", "b": "b"}
item2 = {"a": "a", "b": "b|b:b"}
fields = ["a", "b"]
signature = "a:a|b:b|b:b"
```

:slightly_smiling_face:

Comment (BurnzZ) [reply tests/test_utils.py]: 😆 

Tempted to go with the route of closer to `json.dumps()` but still faster:

```python
>>> python -m timeit "item = {'a':1,'b':2,'c':3}; import json; hash(json.dumps({k:item.get(k) for k in ['a', 'b', 'c']}))"
200000 loops, best of 5: 1.41 usec per loop

>>> python -m timeit "item = {'a':1,'b':2,'c':3}; hash('|'.join(f'{k}:{item.get(k)}' for k in ['a', 'b', 'c']))"
500000 loops, best of 5: 599 nsec per loop

>>> python -m timeit "item = {'a':1,'b':2,'c':3}; hash(','.join(f'\"{k}\": \"{item.get(k)}\"' for k in ['a', 'b', 'c']))"
500000 loops, best of 5: 649 nsec per loop
```

Comment (Gallaecio) [reply tests/test_utils.py]: I really think the original implementation was OK, it would be extremely unlikely for this to be an issue with real data, and if that ever becomes a concern, we could always allow customizing `|` into something else. But I think it makes sense to have an xfail test, as a “we know this can happen in theory” statement in code.

**Review by wRAR — APPROVED** [review]

File: README.rst [hunk 6]
```diff
@@ -146,20 +146,49 @@ non-universal rules that match the URL, the universal ones are applied.
...
 
-.. code-block:: python
+  .. code-block:: python
 
-    DUD_LOAD_RULE_PATHS = [
-        "/home/user/project/custom_rules1.json",
-    ]
+      DUD_LOAD_RULE_PATHS = [
+          "/home/user/project/custom_rules1.json",
+      ]
```
Comment (PyExplorer) [inline README.rst:158]: Would it be helpful to explain how to use locally defined rules with spiders running in Scrapy Cloud?

Comment (wRAR) [reply README.rst:158]: Yeah, this PR doesn't change this part of the doc except formatting but it's nice to have as a separate PR.

**Review by Gallaecio — APPROVED** [review]

**Review by PyExplorer — APPROVED** [review]

File: README.rst (hunk 6 above)
Comment (BurnzZ) [reply README.rst:158]: Thanks for raising @PyExplorer . Created https://github.com/zytedata/duplicate-url-discarder/issues/19 for tracking.

File: tests/test_utils.py (hunk 5 above)
Comment (BurnzZ) [reply tests/test_utils.py]: Added xfail in https://github.com/zytedata/duplicate-url-discarder/pull/17/commits/c562bc6099dd7c3f2dfd8597ca20d8d161f15632. Cheers again for raising!

**Comment by BurnzZ** [comment]
Merging this one. Thanks for the review folks!
